File: webapp/src/transport/csil/conventions.ts
// Conventions shared by every CSIL transport — see `csil-transport-conventions.md`.
//
// The parts the three transports agree on: the CBOR rules (deterministic
// encoding lives in `cbor.ts`, tag-24 payloads here), the version constant, the
// transport status registry, the max-frame guard, and the small map-navigation
// helpers the decoders share. Transport modules build their envelopes from these
// helpers so the bytes match the conformance vectors regardless of object shape.

import { type Cbor, bytes, decode, encode, tag, text, int } from "./cbor.ts";

// Current transport version. A new value is minted only for a breaking change to
// envelope layout or semantics.
export const VERSION = 1;

// CBOR semantic tag wrapping an embedded, opaque CBOR data item (RFC 8949 §3.4.5.1).
export const TAG_ENCODED_CBOR = 24;

// Reserved service ordinal for the transport control plane (Events lifecycle).
export const CONTROL_SERVICE_ORD = 0;

// Default max encoded envelope size for stream/message carriers (16 MiB). A
// carrier rejects anything larger before allocating for it.
export const MAX_FRAME_DEFAULT = 16 * 1024 * 1024;

// Largest max-frame limit a host may configure (2 GiB - 1). The 4-byte length
// prefix can express more, but the limit lives in a signed 32-bit integer in
// several supported languages, so this is the portable ceiling every CSIL
// transport agrees on. Anything above it would also disable the receive guard
// outright, since no wire length could exceed it.
export const MAX_FRAME_LIMIT = 2147483647;

// Check a host-supplied max-frame limit against the conventions doc range
// (1..=MAX_FRAME_LIMIT), so an unusable framer fails where the host configures it
// rather than on its first frame.
export function validateMaxFrame(maxFrame: number): number {
  if (!Number.isInteger(maxFrame) || maxFrame < 1 || maxFrame > MAX_FRAME_LIMIT) {
    throw TransportError.invalidMaxFrame(maxFrame, MAX_FRAME_LIMIT);
  }
  return maxFrame;
}

// Transport-level status. Distinct from application errors, which ride inside the
// payload as a declared `/ ErrorType` arm. See the conventions doc registry.
// Modeled as a numeric code (with named constants) rather than a closed enum so
// host-defined extension codes (>= 64) and unknown codes pass through unchanged.
export const Status = {
  Ok: 0,
  MalformedEnvelope: 1,
  UnknownServiceOrOp: 2,
  Unauthenticated: 3,
  Forbidden: 4,
  VersionUnsupported: 5,
  Internal: 6,
  Unavailable: 7,
  DeadlineExceeded: 8,
} as const;

const STATUS_NAMES: Record<number, string> = {
  0: "ok",
  1: "malformed-envelope",
  2: "unknown-service-or-op",
  3: "unauthenticated",
  4: "forbidden",
  5: "version-unsupported",
  6: "internal",
  7: "unavailable",
  8: "deadline-exceeded",
};

export function statusName(code: number): string {
  return STATUS_NAMES[code] ?? "other";
}

export function statusIsOk(code: number): boolean {
  return code === 0;
}

// The kinds a `TransportError` can take, so callers can branch on the failure
// without string-matching messages.
export type TransportErrorKind =
  | "encode"
  | "decode"
  | "malformed"
  | "frame-too-large"
  | "invalid-max-frame"
  | "unsupported-version"
  | "status"
  | "carrier";

// Errors surfaced by the transport layer (the wire), distinct from a host's
// application errors. A non-zero peer status arrives as `kind: "status"` with the
// registry `code`, so a generated client can route it to its transport-error path.
export class TransportError extends Error {
  readonly kind: TransportErrorKind;
  readonly code?: number;

  constructor(kind: TransportErrorKind, message: string, code?: number) {
    super(message);
    this.name = "TransportError";
    this.kind = kind;
    this.code = code;
  }

  static malformed(message: string): TransportError {
    return new TransportError("malformed", message);
  }

  static frameTooLarge(got: number, max: number): TransportError {
    return new TransportError(
      "frame-too-large",
      `frame of ${got} bytes exceeds max-frame guard of ${max} bytes`,
    );
  }

  static invalidMaxFrame(got: number, limit: number): TransportError {
    return new TransportError(
      "invalid-max-frame",
      `max-frame limit of ${got} is outside the valid range 1..=${limit}`,
    );
  }

  static unsupportedVersion(v: number): TransportError {
    return new TransportError("unsupported-version", `unsupported transport version ${v}`);
  }

  // A non-zero transport status returned by the peer.
  static status(code: number, message?: string): TransportError {
    const name = statusName(code);
    const detail = message ? `: ${message}` : "";
    return new TransportError("status", `transport status ${name} (${code})${detail}`, code);
  }

  static carrier(message: string): TransportError {
    return new TransportError("carrier", `carrier error: ${message}`);
  }
}

// Wrap opaque payload bytes (themselves a CBOR item) in tag 24.
export function tag24(payload: Uint8Array): Cbor {
  return tag(TAG_ENCODED_CBOR, bytes(payload));
}

// Extract the opaque payload bytes from a tag-24 value.
export function untag24(value: Cbor): Uint8Array {
  if (value.t === "tag" && value.tag === TAG_ENCODED_CBOR) {
    if (value.v.t === "bytes") return value.v.v;
    throw TransportError.malformed("tag-24 payload is not a byte string");
  }
  throw TransportError.malformed("expected a tag-24 (encoded-cbor) payload");
}

export { decode as decodeValue, encode as encodeValue };

// Build a deterministically-keyed CBOR map from text-keyed entries. The codec
// sorts at encode time (RFC 8949 §4.2.1), so this just constructs the value; the
// result is byte-identical to the Rust reference's `canon_map`.
export function canonMap(entries: ReadonlyArray<readonly [string, Cbor]>): Cbor {
  return {
    t: "map",
    v: entries.map(([k, v]) => [text(k), v] as const),
  };
}

export function intValue(n: number): Cbor {
  return int(n);
}

export function textValue(s: string): Cbor {
  return text(s);
}

// --- map navigation, mirroring the Rust `map_get` / `get_*` helpers ---

export function mapGet(value: Cbor, key: string): Cbor | undefined {
  if (value.t !== "map") return undefined;
  for (const [k, v] of value.v) {
    if (k.t === "text" && k.v === key) return v;
  }
  return undefined;
}

export function getUint(value: Cbor, key: string): number {
  const f = mapGet(value, key);
  if (f && f.t === "int" && f.v >= 0) return f.v;
  throw TransportError.malformed(`missing or non-integer field '${key}'`);
}

export function getInt(value: Cbor, key: string): number {
  const f = mapGet(value, key);
  if (f && f.t === "int") return f.v;
  throw TransportError.malformed(`missing or non-integer field '${key}'`);
}

export function getText(value: Cbor, key: string): string {
  const f = mapGet(value, key);
  if (f && f.t === "text") return f.v;
  throw TransportError.malformed(`missing or non-text field '${key}'`);
}

export function getTextOpt(value: Cbor, key: string): string | undefined {
  const f = mapGet(value, key);
  return f && f.t === "text" ? f.v : undefined;
}

export function getUintOpt(value: Cbor, key: string): number | undefined {
  const f = mapGet(value, key);
  return f && f.t === "int" && f.v >= 0 ? f.v : undefined;
}

// Verify a decoded envelope's version field, throwing a clear error otherwise.
export function checkVersion(v: number): void {
  if (v !== VERSION) throw TransportError.unsupportedVersion(v);
}

// hex <-> bytes helpers (used by the conformance vectors and for debugging).
export function bytesToHex(b: Uint8Array): string {
  let s = "";
  for (const byte of b) s += byte.toString(16).padStart(2, "0");
  return s;
}

export function hexToBytes(hex: string): Uint8Array {
  if (hex.length % 2 !== 0) throw new Error("hex string has odd length");
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = Number.parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

File: api/internal/transport/conventions.go
// Conventions shared by every CSIL transport — see csil-transport-conventions.md.
//
// This file owns the parts the three transports agree on: the version constant,
// the transport status registry, tag-24 payload wrap/unwrap, the max-frame guard,
// and the canonical-CBOR field accessors the envelopes build on so their bytes
// match the conformance vectors regardless of struct layout.
package transport

import "fmt"

// VERSION is the current transport version. A new value is minted only for a
// breaking change to envelope layout or semantics.
const VERSION uint64 = 1

// TagEncodedCBOR is the CBOR semantic tag wrapping an embedded, opaque CBOR data
// item (RFC 8949 §3.4.5.1).
const TagEncodedCBOR uint64 = 24

// ControlServiceOrd is the reserved service ordinal for the transport control
// plane (Events lifecycle).
const ControlServiceOrd uint64 = 0

// MaxFrameDefault is the default max encoded envelope size for stream/message
// carriers (16 MiB). A carrier rejects anything larger before allocating for it.
const MaxFrameDefault = 16 * 1024 * 1024

// MaxFrameLimit is the largest max-frame limit a host may configure (2 GiB - 1).
// The 4-byte length prefix can express more, but the limit lives in a signed
// 32-bit integer in several supported languages, so this is the portable ceiling
// every CSIL transport agrees on. Anything above it would also disable the receive
// guard outright, since no wire length could exceed it.
const MaxFrameLimit = 2147483647

// ValidateMaxFrame checks a host-supplied max-frame limit against the conventions
// doc range (1..=MaxFrameLimit), so an unusable carrier fails at construction
// rather than on its first frame.
func ValidateMaxFrame(maxFrame int) error {
	if maxFrame < 1 || maxFrame > MaxFrameLimit {
		return ErrInvalidMaxFrame{Got: maxFrame, Limit: MaxFrameLimit}
	}
	return nil
}

// Status is a transport-level status. It is distinct from application errors,
// which ride inside the payload as a declared `/ ErrorType` arm. See the
// conventions doc registry. Equality is by the underlying code, so host-defined
// extension codes (>= 64) and unknown codes compare correctly.
type Status struct{ code int64 }

// The registry codes (conventions doc §4).
var (
	StatusOk                 = Status{0}
	StatusMalformedEnvelope  = Status{1}
	StatusUnknownServiceOrOp = Status{2}
	StatusUnauthenticated    = Status{3}
	StatusForbidden          = Status{4}
	StatusVersionUnsupported = Status{5}
	StatusInternal           = Status{6}
	StatusUnavailable        = Status{7}
	StatusDeadlineExceeded   = Status{8}
)

// StatusFromCode maps a wire code onto a Status, preserving host-defined and
// unknown codes verbatim.
func StatusFromCode(code int64) Status { return Status{code} }

// Code returns the wire status code.
func (s Status) Code() int64 { return s.code }

// IsOk reports whether the status indicates a typed reply is present.
func (s Status) IsOk() bool { return s.code == 0 }

// Name returns the registry name for the status, or "other" for codes outside it.
func (s Status) Name() string {
	switch s.code {
	case 0:
		return "ok"
	case 1:
		return "malformed-envelope"
	case 2:
		return "unknown-service-or-op"
	case 3:
		return "unauthenticated"
	case 4:
		return "forbidden"
	case 5:
		return "version-unsupported"
	case 6:
		return "internal"
	case 7:
		return "unavailable"
	case 8:
		return "deadline-exceeded"
	default:
		return "other"
	}
}

// ErrFrameTooLarge is returned when a frame exceeds the max-frame guard; the
// receiver rejects it before allocating for it.
type ErrFrameTooLarge struct {
	Got int
	Max int
}

func (e ErrFrameTooLarge) Error() string {
	return fmt.Sprintf("frame of %d bytes exceeds max-frame guard of %d bytes", e.Got, e.Max)
}

// ErrInvalidMaxFrame is returned when a host configures a max-frame limit outside
// the valid range.
type ErrInvalidMaxFrame struct {
	Got   int
	Limit int
}

func (e ErrInvalidMaxFrame) Error() string {
	return fmt.Sprintf("max-frame limit of %d is outside the valid range 1..=%d", e.Got, e.Limit)
}

// ErrUnsupportedVersion is returned when an envelope's version is not supported.
type ErrUnsupportedVersion struct{ V uint64 }

func (e ErrUnsupportedVersion) Error() string {
	return fmt.Sprintf("unsupported transport version %d", e.V)
}

// StatusError carries a non-zero transport status returned by a peer, distinct
// from a host's application errors.
type StatusError struct {
	Status  string
	Code    int64
	Message string
}

func (e StatusError) Error() string {
	if e.Message != "" {
		return fmt.Sprintf("transport status %s (%d): %s", e.Status, e.Code, e.Message)
	}
	return fmt.Sprintf("transport status %s (%d)", e.Status, e.Code)
}

// malformed builds a malformed-envelope error.
func malformed(format string, args ...any) error {
	return fmt.Errorf("malformed envelope: "+format, args...)
}

// tag24 wraps opaque payload bytes (themselves a CBOR item) in tag 24.
func tag24(payload []byte) cborValue {
	b := make([]byte, len(payload))
	copy(b, payload)
	return cTag{Num: TagEncodedCBOR, Content: cBytes(b)}
}

// untag24 extracts the opaque payload bytes from a tag-24 value.
func untag24(v cborValue) ([]byte, error) {
	t, ok := v.(cTag)
	if !ok || t.Num != TagEncodedCBOR {
		return nil, malformed("expected a tag-24 (encoded-cbor) payload")
	}
	b, ok := t.Content.(cBytes)
	if !ok {
		return nil, malformed("tag-24 payload is not a byte string")
	}
	out := make([]byte, len(b))
	copy(out, b)
	return out, nil
}

// mapGet looks up a text key in a CBOR map value.
func mapGet(v cborValue, key string) (cborValue, bool) {
	m, ok := v.(cMap)
	if !ok {
		return nil, false
	}
	for _, e := range m {
		if t, ok := e.Key.(cText); ok && string(t) == key {
			return e.Val, true
		}
	}
	return nil, false
}

// asU64 reads a non-negative integer from a decoded CBOR integer value.
func asU64(v cborValue) (uint64, bool) {
	switch x := v.(type) {
	case cUint:
		return uint64(x), true
	case cInt:
		if x >= 0 {
			return uint64(x), true
		}
	}
	return 0, false
}

// asI64 reads a signed integer from a decoded CBOR integer value.
func asI64(v cborValue) (int64, bool) {
	switch x := v.(type) {
	case cUint:
		return int64(x), true
	case cInt:
		return int64(x), true
	}
	return 0, false
}

func getUint(m cborValue, key string) (uint64, error) {
	v, ok := mapGet(m, key)
	if !ok {
		return 0, malformed("missing or non-integer field '%s'", key)
	}
	n, ok := asU64(v)
	if !ok {
		return 0, malformed("field '%s' is not a non-negative integer", key)
	}
	return n, nil
}

func getInt(m cborValue, key string) (int64, error) {
	v, ok := mapGet(m, key)
	if !ok {
		return 0, malformed("missing or non-integer field '%s'", key)
	}
	n, ok := asI64(v)
	if !ok {
		return 0, malformed("missing or non-integer field '%s'", key)
	}
	return n, nil
}

func getText(m cborValue, key string) (string, error) {
	v, ok := mapGet(m, key)
	if !ok {
		return "", malformed("missing or non-text field '%s'", key)
	}
	t, ok := v.(cText)
	if !ok {
		return "", malformed("missing or non-text field '%s'", key)
	}
	return string(t), nil
}

func getTextOpt(m cborValue, key string) *string {
	v, ok := mapGet(m, key)
	if !ok {
		return nil
	}
	if t, ok := v.(cText); ok {
		s := string(t)
		return &s
	}
	return nil
}

func getUintOpt(m cborValue, key string) *uint64 {
	v, ok := mapGet(m, key)
	if !ok {
		return nil
	}
	if n, ok := asU64(v); ok {
		return &n
	}
	return nil
}

// checkVersion verifies a decoded envelope's version field, returning a clear
// error otherwise so an unknown version is never silently misparsed.
func checkVersion(v uint64) error {
	if v == VERSION {
		return nil
	}
	return ErrUnsupportedVersion{V: v}
}
